### ISSUE TITLE
CheckAuth codeFlowCodeRequest incorrect state and suddenly not authenticated

### ISSUE THREAD

**Issue by sweideh** [damienbod/angular-auth-oidc-client#1012]
**Describe the bug**
TLDR; Unauthorized after 5 minutes, is checkAuth using the idToken?
Edit: In conclusion we need to raise our identity token lifetime to stay longer than 5 minutes authenticated. Or is there any other way to verify that a user is authenticated, e.g. by access token?

Hey,

in our auth service we are using the method `checkAuth` to verify if the user is logged in:

```ts
@Injectable()
export class AuthService {
    public get isLoggedIn(): Observable<boolean> {
        return this._isLoggedIn;
    }

    private _isLoggedIn: Observable<boolean>;

    constructor(
        private readonly _oidcSecurityService: OidcSecurityService,
        private readonly _router: Router
    ) {
        this._isLoggedIn = _oidcSecurityService.checkAuth();
    }
```

In our AuthGuard we also using `checkAuth` to verify a logged in user to allow navigation to secured sites.
```ts
import { Injectable } from "@angular/core";
import {
    CanActivate,
    ActivatedRouteSnapshot,
    RouterStateSnapshot,
    Router,
} from "@angular/router";
import { OidcSecurityService } from "angular-auth-oidc-client";
import { Observable } from "rxjs";
import { tap } from "rxjs/operators";

@Injectable()
export class AuthGuard implements CanActivate {
    private readonly _unauthorizedRoute: string;

    constructor(
        private readonly _oidcSecurityService: OidcSecurityService,
        private readonly _router: Router
    ) {
        this._unauthorizedRoute = this._oidcSecurityService.configuration.configuration.unauthorizedRoute;
    }

    public canActivate(
        route: ActivatedRouteSnapshot,
        state: RouterStateSnapshot
    ): Observable<boolean> {
        return this._oidcSecurityService.checkAuth().pipe(
            tap((isAuthenticated) => {
                if (!isAuthenticated) {
                    this._router.navigate([
                        this._unauthorizedRoute ? this._unauthorizedRoute : "/",
                    ]);
                }
            })
        );
    }
}
```

If the user is unauthorized he will be redirected to the root aka login site, as you can see. That happens around 5 minutes. We already checked the access token lifetime. The lifetime is 3600 seconds. `checkAuth` returns false, i checked that and only false.

Now if we try to login again, we don't need to type in the credentials again. The code flow runs smoothly and we are back on our secured site (not login). But there are one warning and one error which both are stating "codeFlowCodeRequest incorrect state". Both message are triggered from the `checkAuth` method. So we don't know why the user is nearly exactly after 5 minutes unauthorized. The session storages also tells an expiration time of 3600 seconds.

**To Reproduce**
-

**Expected behavior**
User should be authorized for 3600 seconds and the `checkAuth` shouldnt return false.

**Screenshots**

**Desktop (please complete the following information):**
 - OS: windows
 - Browser chrome
 - Version 89.0.x
- "angular-auth-oidc-client": "^11.6.2"
- "@angular/core": "~11.1.2"
- "rxjs": "~6.6.3",

**Additional context**
i will add if you need

**Comment by FabianGosebrink** [comment]
Hey, thanks for the question.

The `checkAuth()` method should only be called once, then everything should be set (I thought it could be called mutliple times without issues, so I will look into this). Have you tried to use the `isAuthenticated$` observable to check if the user is authenticated? 

Thanks

**Comment by sweideh** [comment]
Hey @FabianGosebrink,

i tried several things and finally got a working version. I want to share it here, maybe you will extract anything useful or anyone else.

First of all the current AuthService:
```ts
import { Injectable } from "@angular/core";
import { OidcSecurityService } from "angular-auth-oidc-client";
import { Observable, ReplaySubject } from "rxjs";
import { first, switchMap, tap } from "rxjs/operators";

@Injectable()
export class AuthService {

    private checkAuthCompleted$ = new ReplaySubject(1)

    public get unauthorizedRoute(): string {
        return this._unauthorizedRoute;
    }

    private _unauthorizedRoute: string;

    public get postLoginRoute(): string {
        return this._postLoginRoute;
    }

    private _postLoginRoute: string;

    constructor(
        private readonly _oidcSecurityService: OidcSecurityService,
    ) {
        this.checkAuth().subscribe();
        this._unauthorizedRoute = _oidcSecurityService.configuration.configuration.unauthorizedRoute;
        this._postLoginRoute = _oidcSecurityService.configuration.configuration.postLoginRoute;
    }

    public get isAuthenticated$(): Observable<boolean> {
        return this.checkAuthCompleted$.pipe(
            first(),
            switchMap((_) => this._oidcSecurityService.isAuthenticated$)
        );
    }

    public checkAuth(): Observable<boolean> {
        return this._oidcSecurityService.checkAuth().pipe(
            tap((isAuthenticated: boolean) => {
            this.checkAuthCompleted$.next()}
            ));
    }

    public login(redirectRoute?: string): void {
        this.write(
            "redirect",
            redirectRoute ? redirectRoute : window.location.pathname
        );
        this._oidcSecurityService.authorize();
    }

    private write(key: string, value: any): void {
        localStorage.setItem(key, JSON.stringify(value));
    }

    public logout(): void {
        this._oidcSecurityService.logoff();
    }
}

```
So if the AuthService gets initialized, like normal App start or refreshing the site. I am starting the `checkAuth()` progress with subscribe. So isAuthenticated would always have a value when someone tries to get the ob from `isAuthenticated$` 


That is our callback site. Everyone from a login try will be redirected to it.
```ts
import { Component } from "@angular/core";
import { Router } from "@angular/router";
import { AuthService } from "../auth.service";

@Component({
    selector: "auth-callback",
    templateUrl: "./auth-callback.component.html",
    styleUrls: ["./auth-callback.component.scss"],
})
export class AuthCallbackComponent {
    constructor(
        private readonly _router: Router,
        private readonly _authService: AuthService
    ) {
        const postLoginRoute = this._authService.postLoginRoute;
        const storedRoute = this.getAndRemoveStoredRedirectUrl();
        this._authService.isAuthenticated$
            .subscribe((isAuthenticated: boolean) => {
                if (isAuthenticated) {
                    this._router.navigate([
                        storedRoute ? storedRoute : postLoginRoute,
                    ]);
                } else {
                    this._router.navigate(["/"]);
                }
            });
    }

    private getAndRemoveStoredRedirectUrl(): string {
        const path = this.read("redirect");
        localStorage.removeItem("redirect");
        return path;
    }

    private read(key: string): any {
        const data = localStorage.getItem(key);
        if (data) {
            return JSON.parse(data);
        }
        return;
    }
}
```

To protect all other sites we are using an AuthGuard.
```ts
import { Injectable } from "@angular/core";
import {
    CanActivate,
    ActivatedRouteSnapshot,
    RouterStateSnapshot,
    Router,
} from "@angular/router";
import { Observable } from "rxjs";
import { map } from "rxjs/operators";
import { AuthService } from "./auth.service";

@Injectable()
export class DatasafeAuthGuard implements CanActivate {
    private readonly _unauthorizedRoute: string;

    constructor(
        private readonly _authService: AuthService,
        private readonly _router: Router
    ) {
        this._unauthorizedRoute = this._authService.unauthorizedRoute;
    }

    /**
     * Checks the authentication state using the oidcSecurityService.
     * When not authorized, uses the unauthorizedRoute route of the security service to redirect.
     * If unauthorizedRoute property is not set, redirects to root path. ("/")
     */
    public canActivate(
        route: ActivatedRouteSnapshot,
        state: RouterStateSnapshot
    ): Observable<boolean> {
        return this._authService.isAuthenticated$.pipe(
            map((isAuthenticated: boolean) => {
                if (!isAuthenticated) {
                    this._router.navigate([
                        this._unauthorizedRoute ? this._unauthorizedRoute : "/",
                    ]);
                    return false;
                }
                return true;
            })
        );
    }
}
```

Thats an example login page. Just redirects if the user is already authenticated
```ts
import { Component } from "@angular/core";
import { AuthService } from "shared";
import { Router } from "@angular/router";
import { Subject } from "rxjs";
import { filter, take, takeUntil, tap } from "rxjs/operators";

@Component({
    selector: "login",
    templateUrl: "./login.component.html",
    styleUrls: ["./login.component.scss"],
})
export class LoginComponent {
    private _destroy$ = new Subject();

    constructor(
        private readonly _authService: AuthService,
        private readonly _router: Router
    ) {}

    public ionViewWillEnter(): void {
        this._authService
            .isAuthenticated$
            .pipe(
                filter((isAuthenticated: boolean) => isAuthenticated)
            )
            .subscribe(() => {
                this._router.navigate(["/", "index"]);
            });
    }

    public login(): void {
        this._authService.login("/index");
    }
}

```
